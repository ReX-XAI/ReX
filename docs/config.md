# Configuration

This page describes the different sections in the ReX config file, `rex.toml`, and their possible values.
TOML is a [minimal configuration file format](https://toml.io/en/).
An example `rex.toml` file that configures a few of the possible parameters might look like this:

```none
[rex]

gpu = false

batch_size = 4

[rex.onnx]

binary_threshold = 0.5

[rex.visual]

progress_bar = true

[causal]

search_limit = 1000

iters = 30

[causal.distribution]

distribution = 'uniform'

[explanation]

[explanation.spatial]

[explanation.multi]

[explanation.evaluation]
```

N.B. that all section/subsection headers must be present even if no values in that section are set to non-default values.

## [rex] section

This section is for options that affect the general features of ReX.

mask_value = 0

Masking value for mutations, can be either an integer, float or one of the following built-in occlusions 'spectral', 'min', 'mean'.
Default: 0.

seed = 42

Random seed, only set for reproducibility.
Default: None.

gpu = true

Whether to use gpu or not.
Default: true.

batch_size = 64

Batch size for the model.
Default: 1.

### [rex.onnx] section

means = [0.485, 0.456, 0.406]

Means for min-max normalization.
Default: None.

stds = [0.229, 0.224, 0.225]

Standard deviations for min-max normalization.
Default: None.

binary_threshold = 0.5

Binary model confidence threshold.
Anything >= threshold will be classified as 1, otherwise 0.
Default: None.

norm
<!-- unclear to me what this is used for -->

Optional.
Default: None.

### [rex.visual] section

This section is for options that control the appearance of the progress bar and output data visualisation.

progress_bar = false

Whether to show progress bar in the terminal.
Defaults: true.

resize = true

Resize the explanation to the size of the original image.
This uses cubic interpolation and will not be as visually accurate as not resizing.
Default: false.

info = false

Include classification and confidence information in title of plot.
Default: true.

raw = false

Produce unvarnished image with actual masking value.
Defaults: false.

colour = 150

Pretty printing colour for explanations.
Defaults: 200.

heatmap = 'coolwarm'

Matplotlib colourscheme for responsibility map plotting.
Default: 'magma'.
See the [Matplotlib documentation](https://matplotlib.org/stable/users/explain/colors/colormaps.html) for a list of all possible colourmaps.

alpha = 0.2

Alpha blend for main image (PIL Image.blend parameter).
Default: 0.2.

grid = false

Overlay a 10*10 grid on an explanation.
Defaults: false.

mark_segments = false

Mark quickshift segmentation on image.
Default: false.

## [causal] section

This sections is for options relevant to the causal responsibility calculations.

tree_depth = 30

Maximum depth of tree.
Default: 10.
Note that search can actually go beyond this number on occasion, as the check only occurs at the end of an iteration.

search_limit = 1000

Limit on the number of combinations to consider *per iteration*.
Optional.
Default: None.
It is **not** the total work done by ReX over all iterations.
Leaving the search limit at none can potentially be very expensive.

iters = 30

Number of times to run the algorithm.
Defaults: 20.
<!-- from code looks like default is actually 30 -->

min_box_size = 10

Minimum child size, in pixels.
Default: 10.
<!-- is this width or area? -->

confidence_filter = 0.5

Remove passing mutants which have a confidence less than `confidence_filter`.
Default: 0.0 (meaning all mutants are considered).

weighted = false

Whether to weight responsibility by prediction confidence.
Default: false.

queue_style = "area"

queue_style = "intersection" | "area" | "all" | "dc"
Default: "area".

queue_len = 1

Maximum number of things to hold in search queue, either an integer or 'all'.
Default: 1.

concentrate = false

Weight responsibility by size and depth of passing partition.
Default: false.

segmentation

Default: false.

### [causal.distribution] section

This section is for options that control the distribution used to split boxes.

distribution = 'uniform'

Distribution for splitting the box.
Default: 'uniform'.
Possible choices are 'uniform' | 'binom' | 'betabinom' | 'adaptive'.

blend = 0.5

Default: 0.0.

dist_args = [1.1, 1.1]

Supplemental arguments for distribution creation, these are ignored if `distribution` does not take any parameters.
Optional.
Default: None.

## [explanation] section

This section is for options related to identifying the explanation(s).

chunk = 10

Iterate through pixel ranking in chunks, defaults to `causal.min_box_size`.

### [explanation.spatial] section

This section is for options used for spatial search for explanations.

initial_radius = 25

Initial search radius.
Default: 25.

radius_eta = 0.2

Increment to change radius.
Default: 0.2.

no_expansions = 50

Number of times to expand before quitting
Default: 1.
<!-- from code looks like default is 50? -->

### [explanation.multi] section

method = 'spotlight'
<!-- from code looks like this is not used currently -->

Multi-explanation method (only spotlight is currently implemented).

spotlights = 10

Number of spotlights to launch.
Default: 10.

spotlight_size = 24

Default size of spotlight.
Default: 20.

spotlight_eta = 0.2

Decrease spotlight by this amount.
Default: 0.2

obj_function = 'mean'

Objective function for spotlight search.
Possible options 'mean' | 'max' | 'min'.

spotlight_step

Default: 5.

### [explanation.evaluation] section

These options are used if ReX is run with the `--analyse` option.

insertion_step = 100

Insertion/deletion curve step size.
Default: 100.
