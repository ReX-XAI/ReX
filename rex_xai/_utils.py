#!/usr/bin/env python3

import importlib.metadata
from itertools import chain, combinations
from enum import Enum
from typing import Tuple, Union, Dict
from numpy.typing import NDArray
import torch as tt
import numpy as np
from skimage.segmentation import mark_boundaries
from rex_xai.logger import logger
from rex_xai.box import Box

Strategy = Enum("Strategy", ["Global", "Spatial", "MultiSpotlight", "Contrastive"])

Queue = Enum("Queue", ["Area", "All", "Intersection", "DC"])

SpatialSearch = Enum("SpatialSearch", ["NotFound", "Found"])


def powerset(r, reverse=True):
    ps = list(chain.from_iterable(combinations(r, lim) for lim in range(1, len(r) + 1)))
    if reverse:
        return reversed(ps)
    else:
        return ps


def clause_area(clause, areas: Dict) -> int:
    tot = 0
    for c in clause:
        tot += areas[c]
    return tot


class ReXError(Exception):
    pass


class ReXTomlError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXTomlError: {self.message}"


class ReXPathError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXPathError: no such file exists at {self.message}"


class ReXScriptError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXScriptError: {self.message}"


class ReXDataError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXDataError: {self.message}"


class ReXMapError(ReXError):
    def __init__(self, message) -> None:
        self.message = message
        super().__init__(self.message)

    def __str__(self) -> str:
        return f"ReXMapError: {self.message}"


def xlogx(ps):
    f = np.vectorize(_xlogx)
    return f(ps)


def _xlogx(p):
    if p == 0.0:
        return 0.0
    else:
        return p * np.log2(p)


def add_boundaries(
    img: Union[NDArray, tt.Tensor], segs: NDArray, colour=None
) -> NDArray:
    if colour is None:
        m = mark_boundaries(img, segs, mode="thick")
    else:
        m = mark_boundaries(img, segs, colour, mode="thick")
    m *= 255  # type: ignore
    m = m.astype(np.uint8)
    return m


def get_device(gpu: bool):
    if tt.backends.mps.is_available() and gpu:
        return tt.device("mps")
    if tt.device("cuda") and gpu:
        return tt.device("cuda")
    if gpu:
        logger.warning("gpu not available")
    return tt.device("cpu")


def get_map_locations(map, reverse=True):
    if isinstance(map, tt.Tensor):
        map = map.detach().cpu().numpy()
    coords = []
    for i, r in enumerate(np.nditer(map)):
        coords.append((r, np.unravel_index(i, map.shape)))
    coords = sorted(coords, reverse=reverse)
    return coords


def set_boolean_mask_value(
    tensor,
    mode,
    order,
    coords: Union[Box, Tuple[NDArray, NDArray]],
    val: bool = True,
):
    if isinstance(coords, Box):
        if mode in ("spectral", "tabular"):
            h = coords.col_start
            w = coords.col_stop
        else:
            h = slice(coords.row_start, coords.row_stop)
            w = slice(coords.col_start, coords.col_stop)
    else:
        h = coords[0]
        w = coords[1]
    # three channels
    if mode == "RGB":
        # (C, H, W)
        if order == "first":
            tensor[:, h, w] = val
        # (H, W, C)
        else:
            tensor[h, w, :] = val
    elif mode == "L":
        if order == "first":
            # (1, H, W)
            tensor[0, h, w] = val
        else:
            tensor[h, w, :] = val
    elif mode in ("spectral", "tabular"):
        if len(tensor.shape) == 1:
            tensor[h:w] = val
        else:
            tensor[0, h:w] = val
    # elif mode == "tabular":

    elif mode == "voxel":
        logger.warning("not yet implemented")


def ff(obj, fmt):
    """
    Like format(obj, fmt), but returns the string 'None' if obj is None.
    See the help for format() to see acceptable values for fmt.
    """
    return "None" if obj is None else format(obj, fmt)


def version():
    return importlib.metadata.version("rex-xai")
