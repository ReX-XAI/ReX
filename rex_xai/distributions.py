"""distributions module"""

from typing import Optional, Tuple
from enum import Enum
import numpy as np
from scipy.stats import binom, betabinom
from rex_xai.logger import logger

Distribution = Enum("Distribution", ["Binomial", "Uniform", "BetaBinomial", "Adaptive"])


def _betabinom2d(height, width, alpha, beta):
    bx = betabinom(width, alpha, beta)
    by = betabinom(height, alpha, beta)

    w = np.array([bx.pmf(i) for i in range(0, width)])  # type: ignore
    h = np.array([by.pmf(i) for i in range(0, height)])  # type: ignore

    w = np.expand_dims(w, axis=0)
    h = np.expand_dims(h, axis=0)

    u = (h.T * w / np.sum(h.T * w)).ravel()
    p = np.random.choice(np.arange(0, len(u)), p=u)
    return p


def _blend(dist, alpha, base):
    pmf = np.array([base.pmf(x) for x in range(0, len(dist))])
    blend = ((1.0 - alpha) * pmf) + (alpha * dist)
    blend /= np.sum(blend)
    return blend


def _2d_adaptive(map, args: Tuple[int, int, int, int], alpha=0.0, base=None) -> int:
    # if the map exists and is not 0.0 everywhere...
    if map is not None and np.max(map) > 0.0:
        s = map[args[0] : args[1], args[2] : args[3]]
        sf = np.ndarray.flatten(s)
        # sf = np.max(sf) - sf
        sf /= np.sum(sf)

        # base = betabinom(0, len(sf), 1.1, 1.1)
        # if base is not None:
        #     sf = _blend(alpha, base)
        pos = np.random.choice(np.arange(0, len(sf)), p=sf)
        return pos

    # if the map is empty or doesn't exist, return uniform
    return np.random.randint(1, (args[1] - args[0]) * (args[3] - args[2]))


def str2distribution(d: str) -> Distribution:
    """converts string into Distribution enum"""
    if d == "binom":
        return Distribution.Binomial
    elif d == "uniform":
        return Distribution.Uniform
    elif d == "betabinom":
        return Distribution.BetaBinomial
    elif d == "adaptive":
        return Distribution.Adaptive
    else:
        logger.warning(
            "Invalid distribution '%s', reverting to default value Distribution.Uniform",
            d,
        )
    return Distribution.Uniform


def random_coords(d: Optional[Distribution], *args, map=None) -> Optional[int]:
    """generates random coordinates given a distribution and args"""

    try:
        if d == Distribution.Adaptive:
            return _2d_adaptive(map, args[0])

        if d == Distribution.Uniform:
            return np.random.randint(1, args[0])  # type: ignore

        if d == Distribution.Binomial:
            start, stop, *dist_args = args[0]
            return binom(stop - start - 1, dist_args).rvs() + start

        if d == Distribution.BetaBinomial:
            return _betabinom2d(args[1], args[2], args[3][0], args[3][1])

        return None
    except ValueError:
        return None
