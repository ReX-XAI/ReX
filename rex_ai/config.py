#!/usr/bin/env python3

"""config management"""

import sys
from typing import List, Optional
from types import ModuleType
import argparse
from enum import Enum
import os
from os.path import exists, expanduser
import importlib.util
import numpy as np
import toml  # type: ignore


from rex_ai.distributions import str2distribution
from rex_ai.prediction import Prediction
from rex_ai.distributions import Distribution

CAUSAL = Enum("CAUSAL", ["Responsibility"])

Strategy = Enum(
    "Strategy", ["Global", "Spatial", "Spotlight", "MultiSpotlight"]
)

Queue = Enum("Queue", ["Area", "All", "Intersection", "DC"])


class Args:
    """args argument object"""

    # pylint: disable=too-many-instance-attributes
    # pylint: disable=too-few-public-methods
    def __init__(self) -> None:
        self.config_location: Optional[str] = None
        # input file
        self.path: Optional[str] = None
        self.model = None
        self.mode = None
        self.shape: None = None
        self.db: Optional[str] = None
        # gpu support
        self.gpu: bool = True
        # for reproducability
        self.seed: Optional[None] = None
        # for custom scripts (when used as cmdline tool)
        self.custom: Optional[ModuleType] = None
        self.custom_location = None
        self.processed = False
        # min-max normalization
        self.means = None
        self.stds = None
        self.binary_threshold = None
        # verbosity
        self.verbosity = 0
        # whether to show progress bar or not
        self.progress = True
        # save explanation to output
        self.output = None
        self.surface: Optional[str] = None
        self.heatmap: Optional[str] = None
        self.info = True
        self.raw: bool = False
        self.colour: int = 200
        self.mark_segments = False
        self.alpha = 0.2
        self.all = False
        self.resize = False
        self.grid = False
        self.heatmap_colours = 'magma'
        # explanation production strategy
        self.strategy: Strategy = Strategy.Spatial
        self.chunk_size = 25
        self.batch: int = 1
        # args for spatial strategy
        self.spatial_radius: int = 25
        self.spatial_eta: float = 0.2
        self.no_expansions = 50
        # spotlight args
        self.spotlights: int = 10
        self.spotlight_size: int = 20
        self.spotlight_eta: float = 0.2
        self.spotlight_step: int = 5
        self.spotlight_objective_function = np.mean
        # analysis
        self.analyze: bool = False
        self.insertion_step = 100

    def __repr__(self) -> str:
        return (
            f"Args <file: {self.path}, model: {self.model}, "
            + f"gpu: {self.gpu}, "
            + f"mode: {self.mode}, "
            + f"progress_bar: {self.progress}, "
            + f"output_file: {self.output}, surface_plot: {self.surface}, "
            + f"heatmap_plot: {self.heatmap}, "
            + f"means: {self.means}, stds: {self.stds}, "
            + f"explanation_strategy: {self.strategy}, "
            + f"chunk size: {self.chunk_size}, "
            + f"spatial_radius: {self.spatial_radius}, "
            + f"spatial_eta: {self.spatial_eta}, seed: {self.seed}, "
            + f"db: {self.db}, "
            + f"custom: {self.custom}, verbosity: {self.verbosity}, "
            + f"spotlights: {self.spotlights}, spotlight_size: {self.spotlight_size}, "
            + f"spotlight_eta: {self.spotlight_eta}, "
            + f"no_expansions: {self.no_expansions}, "
            + f"obj_function: {self.spotlight_objective_function}, "
        )


class CausalArgs(Args):
    """Creates a causal args object"""

    # pylint: disable=too-many-instance-attributes
    # pylint: disable=too-few-public-methods

    def __init__(self) -> None:
        super().__init__()
        self.config_location = None
        self.type = CAUSAL
        self.tree_depth: int = 10
        self.search_limit: Optional[int] = None
        self.mask_value = 0
        self.confidence_filter = 0.0
        self.min_box_size: int = 10
        self.segmentation = False
        self.data_location: Optional[str] = None
        self.target: Optional[Prediction] = None
        self.distribution: Distribution = Distribution.Uniform
        self.distribution_args: Optional[List] = None
        self.blend = 0.0
        self.weighted: bool = False
        self.iters = 30
        self.concentrate = False
        # queue management
        self.queue_len = 1
        self.queue_style = Queue.Area

    def __repr__(self) -> str:
        return (
            "Causal Args <"
            + Args.__repr__(self)
            + f"mask_value: {self.mask_value}, "
            + f"tree_depth: {self.tree_depth}, search_limit: {self.search_limit}, "
            + f"min_box_size: {self.min_box_size}, weighted: {self.weighted}, "
            + f"confidence_filter: {self.confidence_filter}, "
            + f"data_locations: {self.data_location}, distribution: {self.distribution}, "
            + f"distribution_args: {self.distribution_args}, "
            + f"queue_len: {self.queue_len}, queue_style {self.queue_style}, "
            + f"concentrate: {self.concentrate}, "
            + f"iterations: {self.iters}>"
        )


def get_config_file(path):
    """parses toml file into dictionary"""
    try:
        file_args = toml.load(path)
        return file_args
    except Exception:
        return FileNotFoundError


def cmdargs():
    """parses command line flags"""
    parser = argparse.ArgumentParser(
        prog="ReX",
        description="Explaining AI through causal reasoning",
    )
    parser.add_argument(
        "filename",
        help="file to be processed, assumes that file is 3 channel (RGB or BRG)",
    )
    parser.add_argument(
        "--output",
        nargs="?",
        const="show",
        help="show minimal explanation, optionally saved to <OUTPUT>. Requires a PIL compatible file extension",
    )
    parser.add_argument(
        "-c", "--config", type=str, help="config file to use for rex"
    )

    parser.add_argument(
        "--processed",
        action="store_true",
        help="don't perform any processing with rex itself",
    )

    parser.add_argument(
        "--script", type=str, help="custom loading and preprocessing script"
    )

    parser.add_argument(
        "-v",
        "--verbose",
        action="count",
        default=0,
        help="verbosity level, either -v or -vv, or -vvv",
    )
    parser.add_argument(
        "--surface",
        nargs="?",
        const="show",
        help="surface plot, optionally saved to <SURFACE>",
    )
    parser.add_argument(
        "--heatmap",
        nargs="?",
        const="show",
        help="heatmap plot, optionally saved to <HEATMAP>",
    )

    parser.add_argument("--model", type=str, help="model, must be onnx format")

    parser.add_argument(
        "--strategy",
        "-s",
        type=str,
        help="explanation strategy, one of < multi | spatial | linear | spotlight >",
    )
    parser.add_argument(
        "--database",
        "-db",
        type=str,
        help="store output in sqlite database <DATABASE>, creating db if necessary",
    )
    parser.add_argument(
        "--iters",
        type=int,
        help="manually override the number of iterations set in the config file",
    )

    parser.add_argument(
        "--analyze",
        action="store_true",
        help="area, entropy different and insertion/deletion curves",
    )
    parser.add_argument(
        "--analyse",
        action="store_true",
        help="area, entropy different and insertion/deletion curves",
    )

    parser.add_argument(
        "--show-all",
        action="store_true",
        help="produce a complete breakdown of the image",
    )

    parser.add_argument(
        "--mode",
        "-m",
        type=str,
        help="assist ReX with your input type, one of <tabular>, <spectral>, <RGB>, <L>, <voxel>, <audio>",
    )

    args = parser.parse_args()
    return args


def match_strategy(cmd_args):
    """gets explanation extraction strategy"""
    if cmd_args.strategy == "multi":
        return Strategy.MultiSpotlight
    if cmd_args.strategy == "linear" or cmd_args.strategy == "global":
        return Strategy.Global
    if cmd_args.strategy == "spotlight":
        return Strategy.Spotlight
    if cmd_args.strategy == "spatial":
        pass
    return Strategy.Spatial


def match_queue_style(qs: str) -> Queue:
    qs = qs.lower()
    if qs == "all":
        return Queue.All
    if qs == "area":
        return Queue.Area
    if qs == "dc":
        return Queue.DC
    return Queue.Intersection


def get_objective_function(multi_dict):
    """gets objective function for spotlight search"""
    try:
        f = multi_dict["obj_function"]
        if f == "mean":
            return np.mean
        if f == "max":
            return np.max
        if f == "min":
            return np.min
    except KeyError:
        pass
    return np.mean


def shared_args(cmd_args, args: CausalArgs):
    """parses shared args"""
    if cmd_args.config is not None:
        args.config_location = cmd_args.config
    if cmd_args.model is not None:
        args.model = cmd_args.model
    if cmd_args.surface is not None:
        args.surface = cmd_args.surface
    if cmd_args.heatmap is not None:
        args.heatmap = cmd_args.heatmap
    if cmd_args.output is not None:
        args.output = cmd_args.output
    if cmd_args.verbose > 0:
        args.verbosity = cmd_args.verbose
    if cmd_args.database is not None:
        args.db = cmd_args.database
    if cmd_args.mode is not None:
        args.mode = cmd_args.mode

    args.processed = cmd_args.processed


def get_all_args(path=None):
    # pylint: disable=too-many-branches
    # pylint: disable=too-many-statements
    """parses all arguments from config file and command line"""
    cmd_args = cmdargs()

    path = None
    if cmd_args.config is not None:
        path = cmd_args.config
    else:
        conf_home = expanduser("~/.config/rex.toml")
        # search in current directory first
        if exists("rex.toml"):
            print("using rex.toml in current working directory")
            path = "rex.toml"
        # fallback on $HOME/.config on linux/macos
        elif exists(conf_home):
            print(f"using config in {conf_home}")
            path = conf_home

    args = CausalArgs()
    args.config_location = path

    try:
        config_file_args = get_config_file(path)

        causal_dict = config_file_args["causal"]
        # print(causal_dict)
        if "tree_depth" in causal_dict:
            args.tree_depth = causal_dict["tree_depth"]
        if "search_limit" in causal_dict:
            args.search_limit = causal_dict["search_limit"]
        if "weighted" in causal_dict:
            args.weighted = causal_dict["weighted"]
        if "queue_len" in causal_dict:
            ql = causal_dict["queue_len"]
            if ql != "all":
                args.queue_len = causal_dict["queue_len"]
        if "queue_style" in causal_dict:
            args.queue_style = match_queue_style(causal_dict["queue_style"])
        if "iters" in causal_dict:
            args.iters = causal_dict["iters"]
        if "min_box_size" in causal_dict:
            args.min_box_size = causal_dict["min_box_size"]
        if "confidence_filter" in causal_dict:
            args.confidence_filter = causal_dict["confidence_filter"]
        if "segmentation" in causal_dict:
            args.segmentation = causal_dict["segmentation"]
        if "concentrate" in causal_dict:
            args.concentrate = causal_dict["concentrate"]

        dist = causal_dict["distribution"]
        d = dist["distribution"]
        args.distribution = str2distribution(d)
        if "dist_args" in dist:
            args.distribution_args = dist["dist_args"]
        if "blend" in dist:
            b = dist["blend"]
            if b < 0.0 or b > 1.0:
                print("impossible blend value")
                sys.exit(-1)
            args.blend = dist["blend"]

        rex_dict = config_file_args["rex"]
        if "mask_value" in rex_dict:
            args.mask_value = rex_dict["mask_value"]
        if "seed" in rex_dict:
            args.seed = rex_dict["seed"]
        if "gpu" in rex_dict:
            args.gpu = rex_dict["gpu"]
        if "batch_size" in rex_dict:
            args.batch = rex_dict["batch_size"]

        if "onnx" in rex_dict:
            onnx = rex_dict["onnx"]
            if "means" in onnx:
                args.means = onnx["means"]
            if "stds" in onnx:
                args.stds = onnx["stds"]
            if "binary_threshold" in onnx:
                args.binary_threshold = onnx["binary_threshold"]


        if "visual" in rex_dict:
            if "info" in rex_dict["visual"]:
                args.info = rex_dict["visual"]["info"]
            if "colour" in rex_dict["visual"]:
                args.colour = rex_dict["visual"]["colour"]
            if "color" in rex_dict["visual"]:
                args.colour = rex_dict["visual"]["color"]
            if "alpha" in rex_dict["visual"]:
                args.alpha = rex_dict["visual"]["alpha"]
            if "raw" in rex_dict["visual"]:
                args.raw = rex_dict["visual"]["raw"]
            if "resize" in rex_dict["visual"]:
                args.resize = rex_dict["visual"]["resize"]
            if "progress_bar" in rex_dict["visual"]:
                args.progress = rex_dict["visual"]["progress_bar"]
            if "grid" in rex_dict["visual"]:
                args.grid = rex_dict["visual"]["grid"]
            if "mark_segments" in rex_dict["visual"]:
                args.mark_segments = rex_dict["visual"]["mark_segments"]
            if "heatmap" in rex_dict["visual"]:
                args.heatmap_colours = rex_dict["visual"]["heatmap"]

        explain_dict = config_file_args["explanation"]
        if "chunk" in explain_dict:
            args.chunk_size = explain_dict["chunk"]

        # spatial args
        spatial_dict = explain_dict["spatial"]
        if "initial_radius" in spatial_dict:
            args.spatial_radius = spatial_dict["initial_radius"]
        if "radius_eta" in spatial_dict:
            args.spatial_eta = spatial_dict["radius_eta"]
        if "no_expansions" in spatial_dict:
            args.no_expansions = spatial_dict["no_expansions"]

        multi_dict = explain_dict["multi"]
        if "spotlights" in multi_dict:
            args.spotlights = multi_dict["spotlights"]
        if "spotlight_size" in multi_dict:
            args.spotlight_size = multi_dict["spotlight_size"]
        if "spotlight_eta" in multi_dict:
            args.spotlight_eta = multi_dict["spotlight_eta"]
        if "spotlight_step" in multi_dict:
            args.spotlight_step = multi_dict["spotlight_step"]
        args.spotlight_objective_function = get_objective_function(multi_dict)  # type: ignore


        eval_dict = explain_dict["evaluation"]
        if "insertion_step" in eval_dict:
            args.insertion_step = eval_dict["insertion_step"]

    except KeyError as e:
        print(f"key error {e} in {path}, so reverting to default args")

    except TypeError:
        print(
            "could not find a rex.toml, so running with defaults. This might not produce the effect you want..."
        )

    if cmd_args.script is not None:
        try:
            name, _ = os.path.splitext(cmd_args.script)
            spec = importlib.util.spec_from_file_location(name, cmd_args.script)
            script = importlib.util.module_from_spec(spec)  # type: ignore
            try:
                spec.loader.exec_module(script)  # type: ignore
            except Exception as e:
                print(f"failed to load {name} because of {e}")
            args.custom = script
            args.custom_location = cmd_args.script
        except ImportError:
            pass

    if args.distribution == Distribution.Uniform:
        args.distribution_args = None

    args.path = cmd_args.filename

    shared_args(cmd_args, args)

    args.strategy = match_strategy(cmd_args)

    if args.model is None and args.custom_location is None:
        print("either a <model>.onnx or a python file must be provided")
        sys.exit(-1)

    if cmd_args.iters is not None:
        args.iters = cmd_args.iters


    if cmd_args.analyze or cmd_args.analyse:
        args.analyze = True

    if cmd_args.show_all:
        args.all = True

    return args
