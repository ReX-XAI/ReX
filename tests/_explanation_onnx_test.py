import pytest

from rex_xai.config import CausalArgs, Strategy
from rex_xai.explanation import _explanation, get_prediction_func_from_args
from syrupy.extensions.amber.serializer import AmberDataSerializer
from syrupy.filters import props
from syrupy.matchers import path_type


def test__explanation_snapshot(args_onnx, cpu_device, snapshot):
    prediction_func, model_shape = get_prediction_func_from_args(args_onnx)
    exp = _explanation(args_onnx, model_shape, prediction_func, cpu_device, db=None)

    assert (
        exp
        == snapshot(
            exclude=props(
                "obj_function", "spotlight_objective_function", "model"
            ),  # paths that differ between systems, pointers to functions that will differ between runs
            matcher=path_type(
                types=(CausalArgs,),
                replacer=lambda data, _: AmberDataSerializer.object_as_named_tuple(
                    data
                ),  # needed to allow exclude to work for custom classes
            ),
        )
    )


@pytest.mark.parametrize("strategy", ["global", "spatial"])
def test_extract(exp_onnx, strategy, snapshot):
    if strategy == "global":
        strategy = Strategy.Global
    elif strategy == "spatial":
        strategy = Strategy.Spatial
    else:
        raise ValueError("invalid strategy!")
    exp_onnx.extract(strategy)
    
    assert(exp_onnx.final_mask == snapshot)