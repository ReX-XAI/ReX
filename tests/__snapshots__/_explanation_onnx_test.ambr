# serializer version: 1
# name: test__explanation_snapshot
  Explanation(
    args=CausalArgs(
      all=False,
      alpha=0.2,
      analyze=False,
      batch=1,
      binary_threshold=None,
      blend=0.0,
      chunk_size=10,
      colour=200,
      concentrate=False,
      confidence_filter=0.0,
      config_location=None,
      custom=None,
      custom_location=None,
      data_location=None,
      db=None,
      distribution=<Distribution.Uniform: 2>,
      distribution_args=None,
      gpu=False,
      grid=False,
      heatmap=None,
      heatmap_colours='magma',
      info=True,
      insertion_step=100,
      iters=2,
      mark_segments=False,
      mask_value=0,
      max_spotlight_budget=40,
      means=None,
      min_box_size=10,
      mode=None,
      multi_style='composite',
      no_expansions=4,
      norm=255.0,
      normalise_curves=True,
      output=None,
      path='imgs/dog.jpg',
      permitted_overlap=0.0,
      processed=False,
      progress=True,
      queue_len=1,
      queue_style=<Queue.Area: 1>,
      raw=False,
      resize=False,
      search_limit=1000,
      seed=100,
      segmentation=False,
      shape=None,
      spatial_eta=0.2,
      spatial_radius=25,
      spotlight_eta=0.2,
      spotlight_size=20,
      spotlight_step=5,
      spotlights=10,
      stds=None,
      strategy=<Strategy.Spatial: 2>,
      surface=None,
      tree_depth=10,
      verbosity=0,
      weighted=False,
    ),
    data=Data: RGB, ['N', 3, 224, 224], 224, 224, 3, first
  	 Target:FOUND_CLASS: 209, FOUND_CONF: 0.39424, TARGET_CLASS: n/a, TARGET_CONFIDENCE: n/a,
    explanation=tensor([[[0., 0., 0.,  ..., 0., 0., 0.],
           [0., 0., 0.,  ..., 0., 0., 0.],
           [0., 0., 0.,  ..., 0., 0., 0.],
           ...,
           [0., 0., 0.,  ..., 0., 0., 0.],
           [0., 0., 0.,  ..., 0., 0., 0.],
           [0., 0., 0.,  ..., 0., 0., 0.]],
  
          [[0., 0., 0.,  ..., 0., 0., 0.],
           [0., 0., 0.,  ..., 0., 0., 0.],
           [0., 0., 0.,  ..., 0., 0., 0.],
           ...,
           [0., 0., 0.,  ..., 0., 0., 0.],
           [0., 0., 0.,  ..., 0., 0., 0.],
           [0., 0., 0.,  ..., 0., 0., 0.]],
  
          [[0., 0., 0.,  ..., 0., 0., 0.],
           [0., 0., 0.,  ..., 0., 0., 0.],
           [0., 0., 0.,  ..., 0., 0., 0.],
           ...,
           [0., 0., 0.,  ..., 0., 0., 0.],
           [0., 0., 0.,  ..., 0., 0., 0.],
           [0., 0., 0.,  ..., 0., 0., 0.]]]),
    final_mask=array([[[False, False, False, ..., False, False, False],
          [False, False, False, ..., False, False, False],
          [False, False, False, ..., False, False, False],
          ...,
          [False, False, False, ..., False, False, False],
          [False, False, False, ..., False, False, False],
          [False, False, False, ..., False, False, False]],
  
         [[False, False, False, ..., False, False, False],
          [False, False, False, ..., False, False, False],
          [False, False, False, ..., False, False, False],
          ...,
          [False, False, False, ..., False, False, False],
          [False, False, False, ..., False, False, False],
          [False, False, False, ..., False, False, False]],
  
         [[False, False, False, ..., False, False, False],
          [False, False, False, ..., False, False, False],
          [False, False, False, ..., False, False, False],
          ...,
          [False, False, False, ..., False, False, False],
          [False, False, False, ..., False, False, False],
          [False, False, False, ..., False, False, False]]]),
    maps={209: 1},
    run_stats=dict({
      'avg_box_size': 2747.4166666666665,
      'max_depth_reached': 4,
      'total_failing': 169,
      'total_passing': 41,
    }),
    target_map=array([[ 5.3333335,  5.3333335,  5.666667 , ..., 10.166666 , 10.166666 ,
          10.166666 ],
         [ 5.3333335,  5.3333335,  5.666667 , ..., 10.166666 , 10.166666 ,
          10.166666 ],
         [ 5.3333335,  5.3333335,  5.666667 , ..., 10.166666 , 10.166666 ,
          10.166666 ],
         ...,
         [ 2.3333335,  2.3333335,  2.3333335, ...,  2.       ,  2.       ,
           2.       ],
         [ 2.3333335,  2.3333335,  2.3333335, ...,  2.       ,  2.       ,
           2.       ],
         [ 2.3333335,  2.3333335,  2.3333335, ...,  2.       ,  2.       ,
           2.       ]], dtype=float32),
  )
# ---
